src/service/userProgramPreference.js

```diff
@@ -0,0 +1,371 @@
+const messageUtils = require('./messageUtil');
+const Sequelize = require('sequelize');
+const moment = require('moment');
+const responseCode = messageUtils.RESPONSE_CODE;
+const programMessages = messageUtils.PROGRAM;
+const model = require('../models');
+const {
+  forkJoin
+} = require('rxjs');
+const envVariables = require('../envVariables');
+const RedisManager = require('../helpers/redisUtil');
+var async = require('async');
+const redisManager = new RedisManager();
+
+function getPreferences(req, response) {
+  var data = req.body;
+  var rspObj = req.rspObj;
+  rspObj.apiId = 'api.preference.read';
+  rspObj.apiVersion = '1.0'
+
+  if (!data.request || !data.request.program_id || !data.request.user_id) {
+    rspObj.errCode = programMessages.PREFERENCES.READ.MISSING_CODE
+    rspObj.errMsg = programMessages.PREFERENCES.READ.MISSING_MESSAGE
+    rspObj.responseCode = responseCode.CLIENT_ERROR;
+    loggerError('Error due to missing request or user_id or program_id',
+      rspObj.errCode, rspObj.errMsg, rspObj.responseCode, null, req)
+    return response.status(400).send(errorResponse(rspObj))
+  }
+
+  var redisKey = data.request.user_id + ':' + data.request.program_id;
+  redisManager.getData(redisKey, (err, cacheData) => {
+    if (err || !cacheData) {
+      // If redis cache is failed to give the response
+      getPreferencefromtable(data.request.user_id, data.request.program_id).then((result) => {
+        if (!result.error) {
+          setDatainRedis(data.request.user_id, data.request.program_id, result.res, true, req);
+
+          rspObj.result = {
+            'contributor_preference': {},
+            'sourcing_preference': {},
+            'synced': true
+          }
+          if (result.res !== null) {
+            rspObj.result.contributor_preference = result.res.contributor_preference,
+            rspObj.result.sourcing_preference = result.res.sourcing_preference
+          }
+          rspObj.responseCode = 'OK';
+          return response.status(200).send(successResponse(rspObj));
+        } else {
+          rspObj.responseCode = 'ERR_GET_USER_PREFERENCE_FAILED';
+          rspObj.result = result.result;
+          return response.status(400).send(errorResponse(rspObj));
+        }
+      });
+    } else {
+      // sync back cache data to user_preferences table
+      syncCacheToPreferenceTable(data.request.user_id, data.request.program_id, cacheData);
+      rspObj.responseCode = 'OK'
+      rspObj.result = JSON.parse(cacheData);
+      return response.status(200).send(successResponse(rspObj))
+    }
+  });
+}
+
+function getPreferencefromtable(userId, programId) {
+  return new Promise((resolve, reject) => {
+    const retObj = { error: false, res: {} };
+    model.user_program_preference.findOne({
+      where: {
+        user_id: userId,
+        program_id: programId,
+      }
+    }).then(function (res) {
+      retObj.res = res;
+      return resolve(retObj);
+    })
+      .catch(function (err) {
```

Comment:
Don't leave space between then and catch. If the code is node minified, JS might sometimes add a `;` automatically and the catch block will become invalid.